src/libs/Authentication.js

```diff
@@ -0,0 +1,148 @@
+        authToken: parameters.authToken,
+        shouldRetry: false,
+
+        // Force this request to be made because the network queue is paused when re-authentication is happening
+        forceNetworkRequest: true,
+
+        // Add email param so the first Authenticate request is logged on the server w/ this email
+        email: parameters.email,
+    })
+        .then((response) => {
+            // If we didn't get a 200 response from Authenticate we either failed to Authenticate with
+            // an expensify login or the login credentials we created after the initial authentication.
+            // In both cases, we need the user to sign in again with their expensify credentials
+            if (response.jsonCode !== 200) {
+                switch (response.jsonCode) {
+                    case 401:
+                        throw new Error('passwordForm.error.incorrectLoginOrPassword');
+                    case 402:
+                        // If too few characters are passed as the password, the WAF will pass it to the API as an empty
+                        // string, which results in a 402 error from Auth.
+                        if (response.message === '402 Missing partnerUserSecret') {
+                            throw new Error('passwordForm.error.incorrectLoginOrPassword');
+                        }
+                        throw new Error('passwordForm.error.twoFactorAuthenticationEnabled');
+                    case 403:
+                        if (response.message === 'Invalid code') {
+                            throw new Error('passwordForm.error.incorrect2fa');
+                        }
+                        throw new Error('passwordForm.error.invalidLoginOrPassword');
+                    case 404:
+                        throw new Error('passwordForm.error.unableToResetPassword');
+                    case 405:
+                        throw new Error('passwordForm.error.noAccess');
+                    case 413:
+                        throw new Error('passwordForm.error.accountLocked');
+                    default:
+                        throw new Error('passwordForm.error.fallback');
+                }
+            }
+            return response;
+        });
+}
+
+/**
+ * Reauthenticate using the stored credentials and redirect to the sign in page if unable to do so.
+ *
+ * @param {String} [command] command name for logging purposes
+ * @returns {Promise}
+ */
+function reauthenticate(command = '') {
+    // Prevent any more requests from being processed while authentication happens
+    NetworkStore.setIsAuthenticating(true);
+
+    const credentials = NetworkStore.getCredentials();
+    return Authenticate({
+        useExpensifyLogin: false,
+        partnerName: CONFIG.EXPENSIFY.PARTNER_NAME,
+        partnerPassword: CONFIG.EXPENSIFY.PARTNER_PASSWORD,
+        partnerUserID: credentials.autoGeneratedLogin,
+        partnerUserSecret: credentials.autoGeneratedPassword,
```

Comment:
yeah, AFAIK we are not yet using any keychain features. Maybe using IndexedDB is slightly discouraging hackers here? Gonna resolve this though since it's unrelated to these changes.